lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,505 @@
+  void handleRef(Operation *op);
+
+  /// A helper used by handleRef that can also be used to process ports.
+  ArrayAttr handleRefLike(Operation *op, AnnotationSet annotations,
+                          FIRRTLType type);
+
+  // Helper to handle ports of modules that may have Grand Central annotations.
+  ArrayAttr handlePorts(Operation *op);
+
+  // If true, then some error occurred while the visitor was running.  This
+  // indicates that pass failure should occur.
+  bool failed = false;
+
+public:
+  using FIRRTLVisitor<GrandCentralVisitor>::visitDecl;
+
+  /// Visit FModuleOp and FExtModuleOp
+  void visitModule(Operation *op);
+
+  /// Visit ops that can make up an interface element.
+  void visitDecl(RegOp op) { handleRef(op); }
+  void visitDecl(RegResetOp op) { handleRef(op); }
+  void visitDecl(WireOp op) { handleRef(op); }
+  void visitDecl(NodeOp op) { handleRef(op); }
+
+  /// Process all other ops.  Error if any of these ops contain annotations that
+  /// indicate it as being part of an interface.
+  void visitUnhandledDecl(Operation *op);
+
+  /// Returns true if an error condition occurred while visiting ops.
+  bool hasFailed() { return failed; };
+};
+
+} // namespace
+
+void GrandCentralVisitor::visitModule(Operation *op) {
+  mlir::function_like_impl::setAllArgAttrDicts(op, handlePorts(op).getValue());
+
+  if (isa<FModuleOp>(op))
+    for (auto &stmt : op->getRegion(0).front())
+      dispatchVisitor(&stmt);
+}
+
+/// Process all other operations.  This will throw an error if the operation
+/// contains any annotations that indicates that this should be included in an
+/// interface.  Otherwise, this is a valid nop.
+void GrandCentralVisitor::visitUnhandledDecl(Operation *op) {
+  AnnotationSet annotations(op);
+  auto anno = annotations.getAnnotation(
+      "sifive.enterprise.grandcentral.AugmentedGroundType");
+  if (anno) {
+    auto diag =
+        op->emitOpError()
+        << "is marked as a an interface element, but this op or its ports are "
+           "not supposed to be interface elements (Are your annotations "
+           "malformed? Is this a missing feature that should be supported?)";
+    diag.attachNote()
+        << "this annotation marked the op as an interface element: '" << anno
+        << "'";
+    failed = true;
+  }
+}
+
+/// Process annotations associated with an operation and having some type.
+/// Return the annotations with the processed annotations removed.  If all
+/// annotations are removed, this returns an empty ArrayAttr.
+ArrayAttr GrandCentralVisitor::handleRefLike(mlir::Operation *op,
+                                             AnnotationSet annotations,
+                                             FIRRTLType type) {
+  if (annotations.empty())
+    return ArrayAttr::get(context, {});
+
+  llvm::SmallVector<Attribute> unprocessedAnnos;
+
+  for (auto anno : annotations) {
+    if (!anno.isClass("sifive.enterprise.grandcentral.AugmentedGroundType")) {
```

Comment:
This would probably be useful.  The black box reader pass has three annotations all under the same package (Scala's namespace) that are verbosely written out right now.

That said, usually you're always comparing against the full string and have some action you want to take for each.

What I've really wanted is something that's `llvm::StringSwitch`, but takes a function. E.g.:

```c++
AnnotationSwitch<bool>(annotation)
  .Case<"foo.bar.baz">([&](auto a) { processBaz(a); return true; }
  .Case<"foo.bar.qux">([&](auto a) { processQux(a); return true; }
  .Default([&](auto) { return false; })
```

I think there's some C++ sorcery to do templates with strings, but I'm the wrong person to ask.

I've done this with a `switch(StringSwitch(...))`, but always reverted it because the control flow is very unclear.